src/dia_model.cpp

```diff
@@ -0,0 +1,903 @@
+            cur = ggml_mul_mat(ctx, layer->cross_attn_o, cur);
+        }
+
+
+        // if we ever need to support multiple step decoder runs then this reshape will need to be replaced with permutation.
+        cur = ggml_cont_2d(ctx, cur, cur->ne[0], 2);
+        cur = ggml_add(ctx, cur, residual_cross);
+        struct ggml_tensor * residual_mlp = cur;
+
+        cur = dia_layer_norm(ctx, cur, layer->mlp_norm);
+        // mlp
+        {
+            cur = ggml_mul(ctx, ggml_silu(ctx, ggml_mul_mat(ctx, layer->gate, cur)), ggml_mul_mat(ctx, layer->up, cur));
+            cur = ggml_mul_mat(ctx, layer->out, cur);
+        }
+
+        cur = ggml_add(ctx, cur, residual_mlp);
+    }
+
+    cur = dia_layer_norm(ctx, cur, model->decoder->norm);
+    cur = build_dia_head_outputs(ctx, model, cur);
+    return cur;
+}
+
+void dia_runner::tokenize_sentence(std::string sentence, dia_ubatch & batch) {
+    // Dia's tokenization process is unusual. Essentially Dia takes the byte value for each character and uses that as 
+    // a token array. Additionally, because Dia performs a cfg-scale adjustment before sampling tokens, it is necessary to 
+    // generate with a conditioned context (i.e. with the text) and an unconditioned context (i.e. without any text) so that
+    // proper adjustments can be perfored at each generation step. This means that we need to pad the end of our tokens to the 
+    // max context size for both the conditional and unconditional sequence.
+
+    // if the sentence isn't prepended by dialogue start tokens, [S1] or [S2], then append one.
+    sentence = strip(sentence);
+    std::string start = sentence.substr(0, 4);
+    if (start != "[S1]" && start != "[S2]") {
+        sentence = "[S1] " + sentence;
+    }
+    if (sentence[sentence.size() - 1] != '.') {
+        sentence += ".";
+    }
+
+    // [S1] and [S2] are special character sequences that are replaced with the special tokens 0x01 and 0x02 respectively.
+    std::string r1(1, 1);
+    std::string r2(1, 2);
+    while (sentence.find("[S1]") != std::string::npos) {
+        size_t pos = sentence.find("[S1]");
+        sentence.replace(pos, 4, r1);
+    }
+    while (sentence.find("[S2]") != std::string::npos) {
+        size_t pos = sentence.find("[S2]");
+        sentence.replace(pos, 4, r2);
+    }
+
+    if (sentence.size() > model->max_encoder_context_length) {
+        TTS_ABORT("Dia currently only supports a max of %d characters and received an input of %d characters.", model->max_encoder_context_length, sentence.size());
+    }
+    batch.tokens.reserve(model->max_encoder_context_length * 2);
+    for (auto character : sentence) {
+        batch.tokens.push_back((uint32_t) character);
+    }
+    batch.sentence_length = batch.tokens.size();
+    if (batch.sentence_length <= model->max_delay) {
+        fprintf(stdout, "Your prompt has fewer than 16 tokens. Please note that Dia's generation with prompts that are fewer than 16 tokens is highly inconsistent.");
```

Comment:
This isn't printed immediately due to buffering because/and it is missing the `\n`.